[Cleanup] Remove RenderStyle::whiteSpace().

yulunwu2 opened the pull request:
#### 43470dbb41a950390f2d0a7cddcc038c83bf3652
<pre>
[Cleanup] Remove RenderStyle::whiteSpace().
<a href="https://bugs.webkit.org/show_bug.cgi?id=286084">https://bugs.webkit.org/show_bug.cgi?id=286084</a>
&lt;<a href="https://rdar.apple.com/143067379">rdar://143067379</a>&gt;

Reviewed by Alan Baradlay.

The whiteSpace() shorthand does not represent all combinations of the
longhands textWrapMode() and whiteSpaceCollapse(). whiteSpace() handles
unsupported cases by misclassifying them as WhiteSpace::Normal.

This misclassification and has caused undefinied behavior in the past:

<a href="https://rdar.apple.com/141021052">rdar://141021052</a>

This PR removes usage of whiteSpace() in WebCore and
replaces calls to WhiteSpace() with the corresponding
longhand textWrapMode() and whiteSpaceCollapse.

Canonical link: <a href="https://commits.webkit.org/289303@main">https://commits.webkit.org/289303@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb325cf2f706b00b4c2cd22ef0dbe24e5c68fadc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/86056 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5675 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/91067 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36962 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/88101 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5914 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13739 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66770 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24564 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/89059 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4526 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78039 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47079 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4358 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32340 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/36049 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74950 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33198 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92836 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13303 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/9874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75535 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13517 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73899 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74706 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18420 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18924 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17331 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/6152 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13333 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18688 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/13174 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16606 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14960 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2459,7 +2459,7 @@ static bool elementQualifiesForWritingToolsPreservation(Element* element)
 
     // If the element has `whitespace:pre` (except for the aforementioned exceptions), it should be preserved.
-    if (renderer->style().whiteSpace() == WhiteSpace::Pre)
+    if (renderer->style().whiteSpaceCollapse() == WhiteSpaceCollapse::Preserve)
```
please check if this is really just about white-space-collapse behavior (or the combination of white space collapsing and content breaking)

yulunwu2 replied:
I messaged the PR author for this block of code, will update this CL when I hear back and send out for review.

yulunwu2 replied:
On second thought, I shouldn't do that. I'll update the PR to always assume both collapse and wrap behavior are needed to keep this a pure refactor without any behavior changes. Breaking it up will make it easier to revert and leave the removal of wrap check, if needed, to the code owner.

nt1m replied:
Looking a bit about the history, this is about writing tools (notably the ones that allow you to rewrite an email using AI). https://github.com/WebKit/WebKit/commit/4cf517d5ce5eec2f7df269c82e6c6bb24d348265 originally wanted to target `<blockquote>` / `<pre>` HTML elements. I think it was looking for `white-space: pre` to be able to target other tags like `<xmp>`, `<plaintext>`, `<listing>` but quite frankly I have preference for checking for the HTML tags rather than targeting specific styling. The special exceptions are somewhat evidence that targeting widely for `white-space: pre` is hazardous.

It's probably safer to keep the behavior in this PR and rewrite this in a new PR to check for HTML tags rather than `white-space: pre` quite frankly. cc @rr-codes @pxlcoder

alanbaradlay commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -607,7 +607,7 @@ bool RenderStyle::isIdempotentTextAutosizingCandidate(AutosizeStatus status) con
     if (fields.contains(AutosizeStatus::Fields::FixedHeight)) {
         if (fields.contains(AutosizeStatus::Fields::FixedWidth)) {
-            if (whiteSpace() == WhiteSpace::NoWrap) {
+            if (textWrapMode() == TextWrapMode::NoWrap) {
```
same here.

yulunwu2 replied:
Looking at the PR that added this, it seems like if statement was extracted from a decision tree: I'll update this to check both WhiteSpaceCollapse and TextWrapMode. Thanks for catching this!

alanbaradlay commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -889,18 +889,19 @@ inline bool RenderStyle::NonInheritedFlags::hasAnyPublicPseudoStyles() const
-    // Pre and prewrap do not collapse whitespace.
-    return mode != WhiteSpace::Pre && mode != WhiteSpace::PreWrap && mode != WhiteSpace::BreakSpaces;
+    // WhiteSpaceCollapse::Collapse and WhiteSpaceCollapse::PreserveBreaks
+    // collapse whitespace.
```
does this add any useful information to the line below?

yulunwu2 replied:
Nope! Removing.

nt1m commented on Source/WebCore/editing/cocoa/HTMLConverter.mm:
```diff
@@ -2458,8 +2458,10 @@ static bool elementQualifiesForWritingToolsPreservation(Element* element)
-    // If the element has `whitespace:pre` (except for the aforementioned exceptions), it should be preserved.
-    if (renderer->style().whiteSpace() == WhiteSpace::Pre)
+    // If the element has `whitespace::pre` (except for the aforementioned exceptions), it should be preserved.
+    // `whitespace::pre` is a shorthand for WhiteSpaceCollapse::Preserve && TextWrapMode::NoWrap.
```
```suggestion
    // If the element has `white-space: pre` (except for the aforementioned exceptions), it should be preserved.
    // `white-space: pre` is a shorthand for `white-space-collapse: preserve` & `text-wrap-mode: nowrap`.
```

Can you write it in terms of CSS property names? (would also be good to fix `whitespace` above to be `white-space`)

yulunwu2 replied:
That's a great idea. Done!

nt1m commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -1023,10 +1022,10 @@ inline double RenderStyle::logicalAspectRatio() const
 {
-    // Normal and nowrap do not preserve newlines.
-    return mode != WhiteSpace::Normal && mode != WhiteSpace::NoWrap;
+    // Preserve, PreserveBreaks, and BreakSpaces preserve newlines.
```
This comment can honestly be dropped, the code is self-explanatory here.

yulunwu2 replied:
Agreed. Done!

nt1m commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -607,7 +607,7 @@ bool RenderStyle::isIdempotentTextAutosizingCandidate(AutosizeStatus status) con
     if (fields.contains(AutosizeStatus::Fields::FixedHeight)) {
         if (fields.contains(AutosizeStatus::Fields::FixedWidth)) {
-            if (whiteSpace() == WhiteSpace::NoWrap) {
+            if (whiteSpaceCollapse() == WhiteSpaceCollapse::Collapse && textWrapMode() == TextWrapMode::NoWrap) {
```
I don't think the collapsing of whitespace characters is particularly relevant here (maybe of newlines even).

maybe this should be: 

```suggestion
            if (!preserveNewline() && textWrapMode() == TextWrapMode::NoWrap) {
```

Only really makes a difference with `preserve-spaces`, but I think this expresses intent better (and also would handle the unimplemented `discard` value correctly).

I'm not sure if the collapsing of newlines matters here even, but worth checking with @whsieh

yulunwu2 replied:
I think this would be good to land in a separate PR. We'll be able to add a meaningful PR title and make it easier to revert/debug in the future.

yulunwu2 committed:
[Cleanup] Remove RenderStyle::whiteSpace().
https://bugs.webkit.org/show_bug.cgi?id=286084
<rdar://143067379>

Reviewed by Alan Baradlay.

The whiteSpace() shorthand does not represent all combinations of the
longhands textWrapMode() and whiteSpaceCollapse(). whiteSpace() handles
unsupported cases by misclassifying them as WhiteSpace::Normal.

This misclassification and has caused undefinied behavior in the past:

rdar://141021052

This PR removes usage of whiteSpace() in WebCore and
replaces calls to WhiteSpace() with the corresponding
longhand textWrapMode() and whiteSpaceCollapse.

Canonical link: https://commits.webkit.org/289303@main
